---
title: "R4BDS Final Project"
author: "Group 28"
format: 
  revealjs:
    theme: blood
    slide-number: true
    show-slide-number: all
---

### Introduction

::: columns
::: {.column width="50%"}
-   Gene expression dataset (Golub et al.)
-   Acute myeloid leukemia (AML)
-   Acute lymphoblastic leukemia (ALL)
-   Original experiment aimed for Machine Learning approach
-   Aim: Compare the two groups
:::

::: {.column width="50%"}
::: r-stack
![](../results/99_leukemia.jpg){fig-align="left"}
:::
:::
:::

------------------------------------------------------------------------

### Materials and Methods 1

::: container
::: container
In the project we used the *Gene expression dataset (Golub et al.)*

::: columns
::: {.column width="50%"}
-   Dataset content

    -   72 patients
    -   7129 genes
    -   Combined training and test datasets
    -   Normalization
:::

::: {.column width="50%"}
-   Three tables

    -   72 entries, 2 variables\
    -   7129 entries, 70 variables\
    -   7129 entries, 78 variables
:::
:::
:::
:::

------------------------------------------------------------------------

### Materials and Methods 2

::: panel-tabset
#### Steps

1\. Merging\
2. Cleaning\
3. Transforming\
4. Augmenting\
5. Analysis

#### Raw data 1

```{r}
library("tidyverse")
set.seed(420)
raw_data <- read.csv("../_raw/data_set_ALL_AML_train.csv")
raw_data |> sample_n(5) |> select(2:8)
```

#### Raw data 2

```{r}
set.seed(420)
raw_data2 <- read.csv("../_raw/actual.csv")
raw_data2 |> sample_n(5)
```

#### Tidy data

```{r}
set.seed(420)
tidy_data <- read.csv("../data/03_expression_tidy.csv")
tidy_data |> sample_n(5) |> select(1:7)
```
:::

|                                                                                                        |
|-----|
| \### Results - Data overview ::: columns ::: {.column width="30%"}                                     |
| :::                                                                                                    |
| ::: {.column width="70%"} ![](../results/05_heatmap_AML.png){.fragment .fade-in-then-out height="600"} |
| ![](../results/05_heatmap_AML.png){.fragment .fade-in-then-out height="600"} ::: :::                   |

#### Results - Differential expression heatmap {.smaller}

::: columns
::: {.column width="30%"}
::: r-fit-text
-   Linear modelling applied to find change in expression between ALL (0) and AML (1)
-   Using q-value \< 0.05, significantly differentially expressed genes were filtered
-   50 gene names were filtered based on highest and lowest estimated expression change\
-   Heatmap shows normalized expression differences
:::
:::

::: {.column width="70%"}
::: r-stack
![](../results/05_heatmap_AML.png){.fragment .fade-in-then-out height="600"}

![](../results/05_heatmap_ALL.png){.fragment .fade-in-then-out height="600"}
:::
:::
:::

------------------------------------------------------------------------

### Results - PCA {.smaller}

::: columns
::: {.column width="40%"}
::: r-fit-text
-   59 PCs explain 95% of variance in the data. Top 2 - 26%
-   PCA data points stratified by cancer type.
-   PC1 - AML points spread out, ALL points more clustered
-   PC2 - ALL points spread out, AML points more clustered.
:::
:::

::: {.column width="60%"}
::: r-stack
![](../results/05_pc_barplot.png){.fragment .fade-in-then-out width="900" height="550"}

![](../results/05_pca.png){.fragment .fade-in-then-out}
:::
:::
:::

------------------------------------------------------------------------

### Discussion

-   Original study was designed for machine learning

-   We tried to compare the expression levels for different cancer types

-   Heatmap and 3D-PCA analysis let us to make a classification

-   The dataset includes too many genes which is good for machine learning

-   However, for us less genes were enough come up with a meaningful result\
