<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>00. Running all project scripts</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="00_run_all_files/libs/clipboard/clipboard.min.js"></script>
<script src="00_run_all_files/libs/quarto-html/quarto.js"></script>
<script src="00_run_all_files/libs/quarto-html/popper.min.js"></script>
<script src="00_run_all_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="00_run_all_files/libs/quarto-html/anchor.min.js"></script>
<link href="00_run_all_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="00_run_all_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="00_run_all_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="00_run_all_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="00_run_all_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">00. Running all project scripts</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="loading-all-libraries--" class="level2">
<h2 class="anchored" data-anchor-id="loading-all-libraries--">Loading all libraries ——————————————————-</h2>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.4.4     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
</div>
</section>
<section id="loading-custom-functions" class="level2">
<h2 class="anchored" data-anchor-id="loading-custom-functions">99: Loading custom functions ————————————————</h2>
<section id="heat-map-function" class="level3">
<h3 class="anchored" data-anchor-id="heat-map-function">Heat map function</h3>
</section>
<section id="labeller-function-for-facets" class="level3">
<h3 class="anchored" data-anchor-id="labeller-function-for-facets">Labeller function for facets</h3>
</section>
<section id="function-to-find-n-mostleast-expressed-genes" class="level3">
<h3 class="anchored" data-anchor-id="function-to-find-n-mostleast-expressed-genes">Function to find n most/least expressed genes</h3>
</section>
</section>
<section id="loading-data-files" class="level2">
<h2 class="anchored" data-anchor-id="loading-data-files">01: Loading data files ——————————————————</h2>
<section id="source-of-data" class="level3">
<h3 class="anchored" data-anchor-id="source-of-data">Source of data</h3>
<p>The data set used in this project was originally published by <a href="https://www-science-org.proxy.findit.cvt.dk/doi/10.1126/science.286.5439.531#body-ref-RF14">Gloub <em>et al.</em> (1999)</a> and was downloaded from <a href="https://www.kaggle.com/datasets/crawford/gene-expression/" class="uri">https://www.kaggle.com/datasets/crawford/gene-expression/</a>.</p>
</section>
<section id="brief-data-description" class="level3">
<h3 class="anchored" data-anchor-id="brief-data-description">Brief data description</h3>
<p>There are 3 files of interest in <code>../data/</code>:</p>
<ol type="1">
<li><p><code>actual.csv</code>,</p></li>
<li><p><code>data_set_ALL_AML_independent.csv</code>, and</p></li>
<li><p><code>data_set_ALL_AML_train.csv</code>.</p></li>
</ol>
<p>The first one contains (numeric) patient IDs in column ‘patient’ and cancer type (ALL for acute lymphoblastic leukemia, AML for acute myeloid leukemia) in column ‘cancer’. The remaining two contain gene expression data per gene in row, while the columns hold information on each gene’s description, its accession number, patient ID and something denoted gene “call”. The latter refers to whether the authors of the paper deemed the gene present (P), marginal (M), or absent (A) in the sample from each patient.</p>
</section>
<section id="loading-data" class="level3">
<h3 class="anchored" data-anchor-id="loading-data">Loading data</h3>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 72 Columns: 2
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): cancer
dbl (1): patient

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
New names:
Rows: 7129 Columns: 70
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (36): Gene Description, Gene Accession Number, call...4, call...6, call....
dbl (34): 39, 40, 42, 47, 48, 49, 41, 43, 44, 45, 46, 70, 71, 72, 68, 69, 67...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
New names:
Rows: 7129 Columns: 78
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (40): Gene Description, Gene Accession Number, call...4, call...6, call....
dbl (38): 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
</section>
<section id="writing-combined-data" class="level3">
<h3 class="anchored" data-anchor-id="writing-combined-data">Writing combined data</h3>
</section>
</section>
<section id="cleaning-data" class="level2">
<h2 class="anchored" data-anchor-id="cleaning-data">02: Cleaning data ———————————————————–</h2>
<section id="loading-and-cleaning-data" class="level3">
<h3 class="anchored" data-anchor-id="loading-and-cleaning-data">Loading and cleaning data</h3>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>New names:
Rows: 7129 Columns: 146
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," chr
(74): Gene Description, Gene Accession Number, call...4.x, call...6.x, c... dbl
(72): 39, 40, 42, 47, 48, 49, 41, 43, 44, 45, 46, 70, 71, 72, 68, 69, 67...
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
• `call...72` -&gt; `call...140`
• `call...74` -&gt; `call...142`
• `call...76` -&gt; `call...144`
• `call...78` -&gt; `call...146`</code></pre>
</div>
</div>
</section>
<section id="transforming-data" class="level3">
<h3 class="anchored" data-anchor-id="transforming-data">Transforming data</h3>
</section>
<section id="writing-clean-data" class="level3">
<h3 class="anchored" data-anchor-id="writing-clean-data">Writing clean data</h3>
</section>
</section>
<section id="augmenting" class="level2">
<h2 class="anchored" data-anchor-id="augmenting">03: Augmenting ————————————————————–</h2>
<section id="augmenting-expression-data" class="level3">
<h3 class="anchored" data-anchor-id="augmenting-expression-data">Augmenting expression data</h3>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 72 Columns: 2
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): cancer
dbl (1): patient

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 72 Columns: 7130
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (7130): patient_id, X90857_at, Z84722_at, U16954_at, U19517_at, U14575_a...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
</section>
<section id="scaling-standardizing-expression-data" class="level3">
<h3 class="anchored" data-anchor-id="scaling-standardizing-expression-data">Scaling (standardizing) expression data</h3>
</section>
</section>
<section id="describing-data-exploratory-data-analysis" class="level2">
<h2 class="anchored" data-anchor-id="describing-data-exploratory-data-analysis">04: Describing data (exploratory data analysis) —————————–</h2>
<section id="loading-and-cleaning-data-1" class="level3">
<h3 class="anchored" data-anchor-id="loading-and-cleaning-data-1">Loading and cleaning data</h3>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>New names:
Rows: 7129 Columns: 146
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," chr
(74): Gene Description, Gene Accession Number, call...4.x, call...6.x, c... dbl
(72): 39, 40, 42, 47, 48, 49, 41, 43, 44, 45, 46, 70, 71, 72, 68, 69, 67...
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 72 Columns: 2
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," chr
(1): cancer dbl (1): patient
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
• `call...72` -&gt; `call...140`
• `call...74` -&gt; `call...142`
• `call...76` -&gt; `call...144`
• `call...78` -&gt; `call...146`</code></pre>
</div>
</div>
</section>
<section id="basic-data-set-information" class="level3">
<h3 class="anchored" data-anchor-id="basic-data-set-information">Basic data set information</h3>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7129   74</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 74
  gene_accession_number  `39`  `40`  `42`  `47`  `48`  `49`  `41`  `43`  `44`
  &lt;chr&gt;                 &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 X60483_at               115  -150   300    16   123   445    79    55   152
2 D82347_at                82     5    12   -14    54    94    -3   100    42
3 U46499_at                91    70    24    63    70    92    45   165   -87
4 HG3638-HT3993_s_at      275    40    69   144   164    91   -45   -50    56
5 U32986_s_at             748  1184  1312  1071  1526  1779  1255  1081  1151
# ℹ 64 more variables: `45` &lt;dbl&gt;, `46` &lt;dbl&gt;, `70` &lt;dbl&gt;, `71` &lt;dbl&gt;,
#   `72` &lt;dbl&gt;, `68` &lt;dbl&gt;, `69` &lt;dbl&gt;, `67` &lt;dbl&gt;, `55` &lt;dbl&gt;, `56` &lt;dbl&gt;,
#   `59` &lt;dbl&gt;, `52` &lt;dbl&gt;, `53` &lt;dbl&gt;, `51` &lt;dbl&gt;, `50` &lt;dbl&gt;, `54` &lt;dbl&gt;,
#   `57` &lt;dbl&gt;, `58` &lt;dbl&gt;, `60` &lt;dbl&gt;, `61` &lt;dbl&gt;, `65` &lt;dbl&gt;, `66` &lt;dbl&gt;,
#   `63` &lt;dbl&gt;, `64` &lt;dbl&gt;, `62` &lt;dbl&gt;, `1` &lt;dbl&gt;, `2` &lt;dbl&gt;, `3` &lt;dbl&gt;,
#   `4` &lt;dbl&gt;, `5` &lt;dbl&gt;, `6` &lt;dbl&gt;, `7` &lt;dbl&gt;, `8` &lt;dbl&gt;, `9` &lt;dbl&gt;,
#   `10` &lt;dbl&gt;, `11` &lt;dbl&gt;, `12` &lt;dbl&gt;, `13` &lt;dbl&gt;, `14` &lt;dbl&gt;, `15` &lt;dbl&gt;, …</code></pre>
</div>
</div>
</section>
<section id="transforming-data-1" class="level3">
<h3 class="anchored" data-anchor-id="transforming-data-1">Transforming data</h3>
</section>
<section id="exploratory-data-analysis" class="level3">
<h3 class="anchored" data-anchor-id="exploratory-data-analysis">Exploratory data analysis</h3>
<section id="cancer-sample-sizes" class="level4">
<h4 class="anchored" data-anchor-id="cancer-sample-sizes">Cancer sample sizes</h4>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$x
[1] "Cancer type"

$y
[1] "Count"

$title
[1] "Sample sizes per cancer type"

attr(,"class")
[1] "labels"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Saving 7 x 5 in image</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="distributions-of-mean-expressions" class="level4">
<h4 class="anchored" data-anchor-id="distributions-of-mean-expressions">Distributions of mean expressions</h4>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'cancer'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `labeller` API has been updated. Labellers taking `variable` and `value`
arguments are now deprecated.
ℹ See labellers documentation.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Saving 7 x 5 in image</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="zooming-in-on-around-the-mode" class="level4">
<h4 class="anchored" data-anchor-id="zooming-in-on-around-the-mode">Zooming in on around the MODE</h4>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Saving 7 x 5 in image</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="zooming-in-on-negative-means" class="level4">
<h4 class="anchored" data-anchor-id="zooming-in-on-negative-means">Zooming in on NEGATIVE means</h4>
<div class="cell">
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="zooming-in-on-positive-means" class="level4">
<h4 class="anchored" data-anchor-id="zooming-in-on-positive-means">Zooming in on POSITIVE means</h4>
<div class="cell">
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="finding-highest-and-lowest-expressed-genes" class="level3">
<h3 class="anchored" data-anchor-id="finding-highest-and-lowest-expressed-genes">Finding highest and lowest expressed genes</h3>
</section>
<section id="plotting-distributions-of-10-highest-expressed-genes" class="level3">
<h3 class="anchored" data-anchor-id="plotting-distributions-of-10-highest-expressed-genes">Plotting distributions of 10 highest expressed genes</h3>
<div class="cell" data-fig.asp="0.5">
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 1960</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/highest-ridges-1.png" class="img-fluid" width="600"></p>
</div>
</div>
</section>
<section id="plotting-distribution-of-10-lowest-expressed-genes" class="level3">
<h3 class="anchored" data-anchor-id="plotting-distribution-of-10-lowest-expressed-genes">Plotting distribution of 10 lowest expressed genes</h3>
<div class="cell" data-fig.asp="0.5">
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 592</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/lowest-plots-1.png" class="img-fluid" width="600"></p>
</div>
</div>
</section>
</section>
<section id="analysis--" class="level2">
<h2 class="anchored" data-anchor-id="analysis--">05: Analysis —————————————————————-</h2>
<section id="heat-maps" class="level3">
<h3 class="anchored" data-anchor-id="heat-maps">Heat maps</h3>
<section id="loading-unscaled-data" class="level4">
<h4 class="anchored" data-anchor-id="loading-unscaled-data">Loading unscaled data</h4>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 72 Columns: 7132
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr    (1): cancer
dbl (7131): patient_id, cancer_type_number, X90857_at, Z84722_at, U16954_at,...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Saving 7 x 5 in image</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="loading-scaled-data" class="level4">
<h4 class="anchored" data-anchor-id="loading-scaled-data">Loading scaled data</h4>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 72 Columns: 7132
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr    (1): cancer
dbl (7131): patient_id, cancer_type_number, X90857_at, Z84722_at, U16954_at,...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Saving 7 x 5 in image</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="transforming-data-2" class="level4">
<h4 class="anchored" data-anchor-id="transforming-data-2">Transforming data</h4>
</section>
<section id="linear-modelling" class="level4">
<h4 class="anchored" data-anchor-id="linear-modelling">Linear modelling</h4>
</section>
<section id="processing-the-results-of-modelling" class="level4">
<h4 class="anchored" data-anchor-id="processing-the-results-of-modelling">Processing the results of modelling</h4>
</section>
<section id="making-plots" class="level4">
<h4 class="anchored" data-anchor-id="making-plots">Making plots</h4>
<div class="cell">
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Saving 7 x 5 in image
Saving 7 x 5 in image</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-34-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="pca" class="level3">
<h3 class="anchored" data-anchor-id="pca">PCA</h3>
<section id="loading-the-data" class="level4">
<h4 class="anchored" data-anchor-id="loading-the-data">Loading the data</h4>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 72 Columns: 7132
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr    (1): cancer
dbl (7131): patient_id, cancer_type_number, X90857_at, Z84722_at, U16954_at,...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
</section>
<section id="performing-a-pca-analysis" class="level4">
<h4 class="anchored" data-anchor-id="performing-a-pca-analysis">Performing a PCA analysis</h4>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Saving 7 x 5 in image</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="exploring-variance-explained-by-pcs-and-eigenvalues" class="level4">
<h4 class="anchored" data-anchor-id="exploring-variance-explained-by-pcs-and-eigenvalues">Exploring variance explained by PCs and eigenvalues</h4>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Saving 7 x 5 in image</code></pre>
</div>
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img src="00_run_all_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>